File: dynamic_programming/04_pascal_triangleII/pascal_triangleII.py
from typing import List


class Solution:
    def get_row(self, row_index: int) -> List[int]:
        if not self._check_constraints(row_index):
            raise ValueError('Constraints violated')
        dp = [1]
        if row_index > 0:
            dp.append(1)
            for j in range(1, row_index):
                for i in range(j, 0, -1):
                    dp[i] += dp[i - 1]
                dp.append(1)

        return dp

    def _check_constraints(self, n: int) -> bool:
        if n < 0 or n > 33:
            return False
        return True

File: dynamic_programming/04_pascal_triangleII/pascal_triangleII.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <cstdint>
#include <vector>

using namespace std;


class Solution {
public:
    vector<int> getRow(int rowIndex) {
        if (!checkConstraints(rowIndex)) {
            throw std::runtime_error("Constraints violated");
        }
        vector<int> dp(1, 1);
        if (rowIndex > 0) {
            dp.push_back(1);
            for (int j = 1; j < rowIndex; j++) {
                for (int i = j; i > 0; i--) {
                    dp[i] += dp[i - 1];
                }
                dp.push_back(1);
            }
        }
        return dp;
    }

private:
    bool checkConstraints(int n) {
        if (n < 0 || n > 33) return false;
        return true;
    }
};
